lecture-2023-10-31-xchat/server/src/web.rs

```diff
@@ -0,0 +1,195 @@
+
+    let state = Arc::new(AppState {
+        db_pool: pool,
+        host: address.clone(),
+    });
+
+    let router = Router::new()
+        .route("/", get(user_list))
+        .route("/delete_user", get(delete_user))
+        .layer(Extension(state));
+
+    let listener = match tokio::net::TcpListener::bind(address).await {
+        Ok(listener) => listener,
+        Err(err) => Err(ServerError::WebServerError(err.to_string()))?,
+    };
+
+    match axum::serve(listener, router.into_make_service()).await {
+        Ok(_) => Ok(()),
+        Err(err) => Err(ServerError::WebServerError(err.to_string())),
+    }
+}
+
+
+#[derive(Deserialize)]
+struct LoginFilterParam {
+    login: Option<String>,
+}
+
+
+/// `user_list` is main endpoint (aka landing site) for the web server.
+async fn user_list(
+    state: Extension<Arc<AppState>>,
+    login_filter: Query<LoginFilterParam>,
+) -> Html<String> {
+    let db_result = fetch_chat_messages(
+        &state.db_pool,
+        &login_filter.login,
+    ).await;
+    if let Err(_) = db_result {
+        return Html("Failed to fetch user list!".to_string())
+    }
+    let chat_messages = db_result.unwrap();
+
+    let db_result = fetch_users(&state.db_pool).await;
+    if let Err(_) = db_result {
+        return Html("Failed to fetch user list!".to_string())
+    }
+    let users = db_result.unwrap();
+
+    // Preparing a links for filtering by login and user deletion.
+    let mut filter_links: Vec<String> = vec![
+        format!("<p>Login filter: <a href='http://{}/'>all</a>", state.host),
+    ];
+    let mut delete_links: Vec<String> = vec![
+        "<p>Delete user: ".to_uppercase(),
+    ];
+    for user in users {
+        filter_links.push(format!(
+            ", <a href='http://{}/?login={}'>{}</a>",
+            state.host,
+            user.login,  // TODO: URL-safe encoding
+            user.login,
+        ));
+
+        delete_links.push(format!(
+            ", <a href='http://{}/delete_user?id={}&login={}'>{}</a>",
+            state.host,
+            user.id,    // TODO: perhaps base64 encoding to obfuscate it a bit...
+            user.login, // TODO: URL-safe encoding
+            user.login,
+        ));
+    }
+    let mut filter_links_html = concat(&filter_links);
+    filter_links_html.push_str("</p>");
+    let mut delete_links_html = concat(&delete_links);
+    delete_links_html.push_str("</p>");
+
+    // Construction of the top-level page layout.
+    let mut page: Vec<String> = vec![
+        filter_links_html,
+        delete_links_html,
+        "<table>".to_string(),
+        " <tr>".to_string(),
+        "  <th>".to_string(),
+        "   timestamp".to_string(),
+        "  </th>".to_string(),
+        "  <th>".to_string(),
+        "   user".to_string(),
+        "  </th>".to_string(),
+        "  <th>".to_string(),
+        "   message".to_string(),
+        "  </th>".to_string(),
+        " </tr>".to_string(),
+    ];
```

Comment:
You could probably map over the string literals with an iterator to avoid typing .to_string() all the time :sweat_smile: